File: packages/server/ExchangeGame/GameObjects/Call.cs
using Microsoft.Extensions.Logging;
using System;
using System.Timers;

namespace ExchangeGame
{
    public delegate void OnCompleteDel(Call call);

    public delegate void OnTimeoutDel(Call call);
    public class Call : GameObject
    {
        public Attendee Sender { get; private set; }

        public Attendee Recipient { get; private set; }

        public int Duration { get; private set; }

        public long StartTimestamp { get; private set; }

        private bool _senderConnected = false;

        private bool _recipientConnected = false;

        public OnCompleteDel OnComplete;

        public OnTimeoutDel OnTimeout;

        // TODO configure this
        public int Score { get; set; }

        private const int START_DELAY = 500;

        private bool CallComplete { get { return _senderConnected && _recipientConnected; } }

        private Timer _timer;

        public Call(string displayName, Attendee sender, Attendee recipient): base(displayName)
        {
            Sender = sender;
            Recipient = recipient;
            var rand = new Random();
            // TODO configure this
            Duration = rand.Next(5000, 10000);

            sender.InCall = true;
            recipient.InCall = true;


            SetTimer();
            _logger.LogInformation($"Call {displayName} between {Sender.DisplayName} and {Recipient.DisplayName}, duration {Duration}");
        }

        private void SetTimer()
        {
            _timer = new Timer(Duration + START_DELAY);
            _timer.Elapsed += (Object source, ElapsedEventArgs e) =>
            {
                _timer.Stop();
                Sender.InCall = false;
                Recipient.InCall = false;
                OnTimeout(this);
                _logger.LogInformation($"Call {DisplayName} has elapsed");
            };

            DateTimeOffset now = (DateTimeOffset)DateTime.UtcNow;
            StartTimestamp = now.ToUnixTimeMilliseconds() + Convert.ToInt64(START_DELAY);
            _timer.Start();
        }

        public bool TryConnect(Attendee attendee)
        {
            var success = false;

            if (Sender == attendee)
            {
                success = true;
                _senderConnected = true;
            }

            if (Recipient == attendee)
            {
                success = true;
                _recipientConnected = true;
            }

            if (CallComplete)
            {
                _logger.LogInformation($"Complete call: {DisplayName}");
                Sender.InCall = false;
                Recipient.InCall = false;
                OnComplete(this);
            }

            return success;
        }
    }
}

File: packages/server-v2/src/GameObjects/Call.ts
import Attendee from "./Attendee";
import { generateId, getRandomInt } from "../utils";

export default class Call {
  public id: number;
  public displayName: string;
  public sender: Attendee;
  public recipient: Attendee;
  public duration: number;
  public startTimestamp: number;
  public onComplete: ((call: Call) => void)[] = [];
  public onTimeout: ((call: Call) => void)[] = [];
  public score: number = 15;
  private _senderConnected: boolean = false;
  private _recipientConnected: boolean = false;
  private START_DELAY = 500;
  private timer: NodeJS.Timeout;

  constructor(displayName: string, sender: Attendee, recipient: Attendee) {
    this.displayName = displayName;
    this.id = generateId();
    this.sender = sender;
    this.recipient = recipient;
    this.duration = getRandomInt(15, 30);
    sender.inCall = true;
    recipient.inCall = true;

    this.startTimestamp = Date.now();
    this.timer = setTimeout(() => {
       this.onTimeout.forEach(cb => cb(this));
    }, this.duration + this.START_DELAY);

    console.log(
      `Call ${displayName} between ${this.sender.displayName} and ${this.recipient.displayName} duration ${this.duration}`
    );
  }

  public tryConnect(attendee: Attendee): boolean {
    let sucess = false;
    if (this.sender === attendee) {
      sucess = true;
      this._senderConnected = true;
    }

    if (this.recipient === attendee) {
      sucess = true;
      this._recipientConnected = true;
    }

    if (this.callComplete) {
      clearTimeout(this.timer);
      console.log(`Complete call: ${this.displayName}`);
      this.sender.inCall = false;
      this.recipient.inCall = false;
      this.onComplete.forEach(cb => cb(this));
    }

    return sucess;
  }

  private get callComplete() {
    return this._senderConnected && this._recipientConnected;
  }
}
